secret/secret.go

```diff
@@ -18,190 +17,222 @@ import (
-	secret, err = s.Get(SECRET_NAME, metav1.GetOptions{})
-	if err == nil {
-		return secret
-	}
-
-	// (4) Neither R-1 nor unversioned available.
-	return nil
+	return configMap
 }
 
-func CreateSecrets(s secretInterface) (secrets, updates *v1.Secret) {
-	releaseRevision := getEnv("RELEASE_REVISION")
-
-	if releaseRevision == "" {
-		logFatal("RELEASE_REVISION is missing or empty.")
-		return nil, nil
-	}
+func GetSecrets(s secretInterface, configMap *v1.ConfigMap) *v1.Secret {
+	currentName := configMap.Data[CURRENT_SECRETS_NAME]
 
-	rv, err := strconv.Atoi(releaseRevision)
-	if err != nil {
-		logFatal(err)
-		return nil, nil
+	newName := getEnv("KUBE_SECRETS_GENERATION_NAME")
+	if newName == "" {
+		logFatal("KUBE_SECRETS_GENERATION_NAME is missing or empty.")
+		return nil
 	}
-
-	secretUpdateName := fmt.Sprintf("%s-%s", SECRET_UPDATE_NAME, releaseRevision)
-	secretName := fmt.Sprintf("%s-%s", SECRET_NAME, releaseRevision)
-
-	log.Printf("Checking for chart-provided `%s`\n", secretUpdateName)
-
-	// secret updates *must* exist
-	updates, err = s.Get(secretUpdateName, metav1.GetOptions{})
-	if err != nil {
-		logFatal(err)
-		return nil, nil
+	if newName == currentName {
+		log.Printf("Secret `%s` already exists; nothing to do\n", newName)
+		return nil
 	}
 
-	// We always create a new secret
-	secrets = &v1.Secret{
+	newSecrets := &v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: secretName,
+			Name: newName,
 		},
 		Data: map[string][]byte{},
 	}
 
-	log.Println("Checking for previous secret")
-
-	// Check for existing secret to use as baseline
-	previousSecrets := FindPreviousSecret(s, rv)
-	if previousSecrets != nil {
-		log.Printf("Importing previous secret `%s`\n", previousSecrets.Name)
-		secrets.Data = previousSecrets.Data
+	currentSecrets, err := s.Get(currentName, metav1.GetOptions{})
+	if err == nil {
+		newSecrets.Data = currentSecrets.Data
+	} else {
+		if currentName != LEGACY_SECRETS_NAME {
+			logFatal(fmt.Sprintf("Cannot get previous version of secrets using name '%s'.", currentName))
+			return nil
+		}
+		// This is a new installation, so make sure the configmap is created, not updated
+		configMap.Data[CURRENT_SECRETS_NAME] = ""
 	}
 
-	log.Printf("Fill secret `%s`\n", secrets.Name)
-	return
+	return newSecrets
 }
 
-func GenerateSecrets(manifest model.Manifest, secrets, updates *v1.Secret) {
+func GenerateSecrets(manifest model.Manifest, secrets *v1.Secret, configMap *v1.ConfigMap) {
+	secretsGeneration := getEnv("KUBE_SECRETS_GENERATION_COUNTER")
+	if secretsGeneration == "" {
+		logFatal("KUBE_SECRETS_GENERATION_COUNTER is missing or empty.")
+		return
+	}
+
+	if secretsGeneration != configMap.Data[CURRENT_SECRETS_GENERATION] {
+		log.Printf("Rotating secrets; generation '%s' -> '%s'\n", configMap.Data[CURRENT_SECRETS_GENERATION], secretsGeneration)
+		for name := range secrets.Data {
+			rotate := true
+			for _, configVar := range manifest.Configuration.Variables {
```

Comment:
I didn't think the quadratic performance matters because we should never have even 1000 secrets, but your code is obviously better, so I used it (also for the later cleanup).